test/signer.ts

```diff
@@ -0,0 +1,504 @@
+          const error = new Error('Error.');
+          sandbox.stub(authClient, 'sign').rejects(error);
+
+          await assert.rejects(() => signer.getSignedUrl(CONFIG), {
+            name: 'SigningError',
+            message: error.message,
+          });
+        });
+      });
+
+      describe('v2 signed URL', () => {
+        beforeEach(() => (CONFIG.version = 'v2'));
+
+        it('should create a v2 signed url when specified', async () => {
+          const signStub = sandbox
+            .stub(authClient, 'sign')
+            .resolves('signature');
+
+          const EXPECTED_BLOB_TO_SIGN = [
+            'GET',
+            '',
+            '',
+            Math.round(Number(CONFIG.expires) / 1000),
+            `/${bucket.name}/${encodeURIComponent(file.name)}`,
+          ].join('\n');
+
+          const signedUrl = await signer.getSignedUrl(CONFIG);
+
+          assert.strictEqual(typeof signedUrl, 'string');
+          const expires = Math.round(Number(CONFIG.expires) / 1000);
+          const expected =
+            'https://storage.googleapis.com/bucket-name/file-name.png?' +
+            'GoogleAccessId=client-email&Expires=' +
+            expires +
+            '&Signature=signature';
+          assert.strictEqual(signedUrl, expected);
+          assert.deepStrictEqual(
+            signStub.getCall(0).args[0],
+            EXPECTED_BLOB_TO_SIGN
+          );
+        });
+
+        it('should not modify the configuration object', async () => {
+          const originalConfig = Object.assign({}, CONFIG);
+
+          await signer.getSignedUrl(CONFIG);
+          assert.deepStrictEqual(CONFIG, originalConfig);
+        });
+
+        it('should return an error if signBlob errors', async () => {
+          const error = new Error('Error.');
+
+          sandbox.stub(authClient, 'sign').rejects(error);
+
+          await assert.rejects(() => signer.getSignedUrl(CONFIG), {
+            name: 'SigningError',
+            message: error.message,
+          });
+        });
+
+        it('should URI encode file names', async () => {
+          file.name = 'directory/name.png';
+          const signedUrl = await signer.getSignedUrl(CONFIG);
+          assert(signedUrl.includes(file.name));
+        });
+
+        it('should URI encode file name with special characters', async () => {
+          file.name = "special/azAZ!*'()*%/file.jpg";
+          const signedUrl = await signer.getSignedUrl(CONFIG);
+          assert(
+            signedUrl.includes('special/azAZ%21%2A%27%28%29%2A%25/file.jpg')
+          );
+        });
+      });
+
+      describe('cname', () => {
+        it('should use a provided cname', async () => {
+          const host = 'http://www.example.com';
+          const configWithCname = Object.assign({cname: host}, CONFIG);
+
+          const signedUrl = await signer.getSignedUrl(configWithCname);
+          const expires = Math.round(Number(CONFIG.expires) / 1000);
+          const expected =
```

Comment:
Checking against the completed URL means the test from `describe('getSignedUrl')` is actually testing behavior in `getSignedUrlV2`. Can we intercept the call to `getSignedUrlV2(config)` and verify config.cname is set to the user-provided input?

Then the same system throughout the rest of the tests in this file, so we can stay as local as possible to the function we are testing. It looks like we'd want new `describe` blocks for:

- getSignedUrlV2
- getSignedUrlV4
- getCanonicalHeaders
- getCanonicalRequest
- getCanonicalQueryParams
- getResourcePath
- parseExpires

Reply:
Make sense! I refactored all the tests in URLSigner so they are localized. Thanks @stephenplusplus